### PR TITLE
feat: pin table sorting function

### PR DESCRIPTION
**Pull request by marharyta**
# Definition of Done / Review checklist

## Reviewability
- [x] link to issue: https://github.com/CSCfi/rems/issues/2183
- [ ] note if PR is on top of other PR
- [ ] note if related change in rems-deploy repo
- [x] consider adding screenshots for ease of review

## Documentation
- [x] update changelog if necessary
- [ ] add or update docstrings for namespaces and functions
- [ ] components are added to guide page
- [ ] documentation _at least_ for config options (i.e. docs folder)
- [ ] ADR for major architectural decisions or experiments

## Testing
- [x] complex logic is unit tested
- [ ] valuable features are integration / browser / acceptance tested automatically

## Follow-up
- [ ] new tasks are created for pending or remaining tasks
- [ ] no critical TODOs left to implement

### CODE REVIEW

**Comment by marharyta** [comment]
Unit tests not ready yet, but coming

**Comment by marharyta** [comment]
After the initial comments from @Macroz 
[ ] the diff is not clean, there seems to be unrelated changes (more later)
[ ] the commits look a bit strange (e.g. three merges in a row), consider rebasing on top of the current main, after fixing the things
[x]a debug console log is still there - removed
[x] almost duplicate functions in the parse-month-to-number and parse-year-to-number - removed and renamed
[ ] why do the numbers need to have these special parsing functions in the first place? the solution should likely be more generic than just supporting a year and a number (i.e. the real diaarinumero have multiple parts)
[ ] there are changes in table.cljs that I don't understand and that should probably not be there
[ ] unrelated changes in browser tests?
[x] unrelated changes in test-util?
[ ] if/when using the javascript parseInt, always specify a radix of 10 for backwards compatibility
[x] application-util I think is the correct place for the parser 👍

File: src/cljc/rems/common/application_util.cljc [hunk 1]
```diff
@@ -27,3 +29,15 @@
 
 (defn is-handler? [application user]
   (contains? (workflow-handlers application) user))
+
+(defn- str-to-int [s]
+  #?(:clj (Integer/parseInt s)
+     :cljs (js/parseInt s 10)))
+
+(defn parse-sequence-number-within-year-to-number [external-id]
+  (str-to-int (get (str/split external-id #"/") 1)))
```
Comment (Macroz) [inline src/cljc/rems/common/application_util.cljc:38]: This function assumes that the format will always be `year/number` and the external-id could be something like `THL/2492/4.00.00/2020` so this should be written a bit more generic, and the name would then change. I would name it  something like `parse-sortable-external-id`

File: src/cljc/rems/common/application_util.cljc [hunk 2]
```diff
@@ -27,3 +29,15 @@
...
+
+;; put test here
+(deftest test-validate-external-id
```
Comment (Macroz) [inline src/cljc/rems/common/application_util.cljc:41]: The unit test should always be named `test-the-function-name` so here `test-parse-sequence-number-within-year-to-number`.

File: src/cljc/rems/common/application_util.cljc [hunk 3]
```diff
@@ -27,3 +29,15 @@
...
+  (is (= 10 (parse-sequence-number-within-year-to-number "2020/10")))
+  (is (= 6 (parse-sequence-number-within-year-to-number "2020/6"))))
```
Comment (Macroz) [inline src/cljc/rems/common/application_util.cljc:43]: I don't see a case for `nil`, for different years, and different formats like `THL/2492/4.00.00/2020`

File: src/cljs/rems/table.cljs [hunk 4]
```diff
@@ -181,17 +181,21 @@
...
     (rf/subscribe [::sorting table])])
  (fn [[rows sorting] _]
    (->> rows
-        (sort-by #(get-in % [(:sort-column sorting) :sort-value])
-                 (case (:sort-order sorting)
-                   :desc #(compare %2 %1)
-                   #(compare %1 %2))))))
+        (sort-by (fn [row]
+                   (get-in row [(:sort-column sorting) :sort-value]))
+                 #(normal-sort sorting %1 %2)))))
```
Comment (Macroz) [inline src/cljs/rems/table.cljs:198]: Why this change?

File: src/cljs/rems/table.cljs [hunk 5]
```diff
@@ -181,17 +181,21 @@
         ;;       (or then we could just not validate these internal row representations)
         (s/validate Rows))))
 
+(defn- normal-sort [sorting option1 option2]
+  (case (:sort-order sorting)
+    :desc (compare option2 option1)
+    :asc (compare option1 option2)))
+
```
Comment (Macroz) [inline src/cljs/rems/table.cljs:188]: This is not identical to the old code. Why?

File: test/clj/rems/test_browser.clj [hunk 6]
```diff
@@ -906,19 +906,19 @@
...
+                             (q (get-in application [:application/applicant :name]))
+                             (q (text/localize-time (get-in application [:application/first-submitted])))
+                             (q "Applied")
+                             (q "")
+                             (q "Tämä on monimutkainen arvo skandein varusteltuna!")])]
+                 (->> #"applications_.*\.csv"
+                      btu/downloaded-files
+                      first
+                      slurp
+                      str/split-lines)))))))
```
Comment (Macroz) [inline test/clj/rems/test_browser.clj:924]: It's ok to fix some formatting problem but it's likely that there will be conflict with this soon so I'd just remove it because you are not touching the file otherwise.

File: src/cljs/rems/table.cljs (hunk 5 above)
Comment (marharyta) [reply src/cljs/rems/table.cljs:188]: I believed it would look cleaner this way, I might be wrong

File: src/cljs/rems/table.cljs (hunk 4 above)
Comment (marharyta) [reply src/cljs/rems/table.cljs:198]: I believed it would look cleaner this way, I might be wrong

File: src/cljc/rems/common/application_util.cljc (hunk 3 above)
Comment (marharyta) [reply src/cljc/rems/common/application_util.cljc:43]: is there any place I can look at other possible formats?

File: src/cljc/rems/common/application_util.cljc (hunk 2 above)
Comment (marharyta) [reply src/cljc/rems/common/application_util.cljc:41]: ok!

**Comment by marharyta** [comment]
Commit squash in progress.

**Comment by marharyta** [comment]
Update:
After the initial comments from @Macroz
[x] the diff is not clean, there seems to be unrelated changes (more later)
[ ] the commits look a bit strange (e.g. three merges in a row), consider rebasing on top of the current main, after fixing the things
[x]a debug console log is still there - removed
[x] almost duplicate functions in the parse-month-to-number and parse-year-to-number - removed and renamed
[ x] why do the numbers need to have these special parsing functions in the first place? the solution should likely be more generic than just supporting a year and a number (i.e. the real diaarinumero have multiple parts)
[x] there are changes in table.cljs that I don't understand and that should probably not be there - to me adding a separate function to do sorting seemed like a cleaner solution
[x] unrelated changes in browser tests?
[x] unrelated changes in test-util?
[x ] if/when using the javascript parseInt, always specify a radix of 10 for backwards compatibility
[x] application-util I think is the correct place for the parser 👍

File: src/cljc/rems/common/application_util.cljc [hunk 7]
```diff
@@ -27,3 +29,25 @@
...
+
+(defn- printToConsole [s]
+  #?(:clj (println s)
+     :cljs (js/console.log s)))
+
+(defn parse-sortable-external-id [external-id]
+  ;; the idea is to parse all numbers from the string, then add them up to form an external id - 
+  ;; comes from a theory that it is somehow tied to the date or the numbers are created in sequence
+  (cond
+    (if (seq (re-seq #"\d+" external-id)) "true" "false") (reduce + (map (fn [str] (str-to-int str)) (re-seq #"\d+" external-id)))
```
Comment (opqdonut) [inline src/cljc/rems/common/application_util.cljc:45]: Why sum the numbers? This leads to funny sort orders, for example:

- ABC/2000.1
- ABC/2003.1
- ABC/2001.5

It's of course hard to sort arbitrary identifiers in general, but perhaps sorting the number sequences lexicographically would be a better solution. Luckily clojure vectors work nicely for this:

```
(map #(vec (re-seq #"\d+" %)) ["ABC/2000.1" "ABC/2003.1" "ABC/2001.5" "ABC/2001.11"])
  ==> (["2000" "1"] ["2003" "1"] ["2001" "5"] ["2001" "11"])
(sort [[2000 1] [2003 1] [2001 5] [2001 11]])
  ==> ([2000 1] [2001 5] [2001 11] [2003 1])
```

In any case, a test that sorts various lists of identifiers would be useful both to document the behaviour and as a place to add regression cases.

Comment (marharyta) [reply src/cljc/rems/common/application_util.cljc:45]: ok, Ill redo it :)

File: CHANGELOG.md [hunk 8]
```diff
@@ -33,6 +33,7 @@ Changes since v2.12
 - OIDC access tokens are now revoked on logout if the OIDC server provides a `revocation_endpoint`. (#2176)
 - Application attachment fields now accept multiple attachments. (#2122)
 - It's now possible to add a text to the login page after the login button using extra translations (:t.login/intro2) (#2214)
+- It is possible to sort applications based on the application external id accumulated number (#2183)
```
Comment (Macroz) [inline CHANGELOG.md:36]: It was possible before? This should say what it actually does (i.e. tries to sort them correctly).

File: src/cljc/rems/common/application_util.cljc [hunk 9]
```diff
@@ -27,3 +29,25 @@
 
 (defn is-handler? [application user]
   (contains? (workflow-handlers application) user))
+
+(defn- str-to-int [s]
+  #?(:clj (Integer/parseInt s)
+     :cljs (js/parseInt s 10)))
+
+(defn- printToConsole [s]
```
Comment (Macroz) [inline src/cljc/rems/common/application_util.cljc:37]: naming-convention-is-kebab-case also this function is not required?

File: src/cljc/rems/common/application_util.cljc (hunk 7 above)
Comment (Macroz) [reply src/cljc/rems/common/application_util.cljc:45]: Doesn't this also lose all the strings for comparison purposes? Also test what happens with 0-prefixed numbers i.e. 001 and stuff like 4.00.00. See `THL/2492/4.00.00/2020`

File: src/cljc/rems/common/application_util.cljc [hunk 10]
```diff
@@ -27,3 +29,25 @@
...
+  ;; the idea is to parse all numbers from the string, then add them up to form an external id - 
+  ;; comes from a theory that it is somehow tied to the date or the numbers are created in sequence
+  (cond
+    (if (seq (re-seq #"\d+" external-id)) "true" "false") (reduce + (map (fn [str] (str-to-int str)) (re-seq #"\d+" external-id)))
+    :else external-id))
+
+;; put test here
+(deftest test-parse-sortable-external-id
+  (is (= 2030 (parse-sortable-external-id "2020/10")))
+  (is (= 2092 (parse-sortable-external-id "THL/54/14.01.00/2002-rems/21")))
```
Comment (Macroz) [inline src/cljc/rems/common/application_util.cljc:51]: This should probably be `["THL/" 54 "/14.01.00/" 2002 "-rems/" 21]`

And THL/2492/4.00.00/2020 would be `["THL/" 2492 "/4.00.00/" 2020]`

File: src/cljs/rems/table.cljs (hunk 5 above)
Comment (Macroz) [reply src/cljs/rems/table.cljs:188]: But you don't handle the default case here like the previous code did.

File: src/cljs/rems/application_list.cljs [hunk 11]
```diff
@@ -81,7 +81,7 @@
    (map (fn [app]
           {:key (:application/id app)
            :id {:value (:application/id app)}
-           :external-id {:value (:application/external-id app)}
+           :external-id {:value (:application/external-id app) :sort-value (application-util/parse-sortable-external-id (:application/external-id app))}
```
Comment (Macroz) [inline src/cljs/rems/application_list.cljs:84]: I think you should split the line at `:sort-value` like in the next line.

File: src/cljc/rems/common/application_util.cljc (hunk 3 above)
Comment (Macroz) [reply src/cljc/rems/common/application_util.cljc]: I searched some with Google and several are in this format.

File: src/cljc/rems/common/application_util.cljc (hunk 7 above)
Comment (opqdonut) [reply src/cljc/rems/common/application_util.cljc:45]: 0-prefixed numbers should work fine when parsed into ints.

File: src/cljc/rems/common/application_util.cljc (hunk 10 above)
Comment (opqdonut) [reply src/cljc/rems/common/application_util.cljc:51]: Why not
```
["THL/" 54 "/" 14 "." 1 "." 0 "/" 2002 "-rems/" 21]
["THL/" 2492 "/" 4 "." 0 "." 0 "/" 2020]
```
more consistent handling of numbers and results in the same ordering (assuming 0-padding is consistent)

File: src/cljc/rems/common/application_util.cljc (hunk 7 above)
Comment (Macroz) [reply src/cljc/rems/common/application_util.cljc:45]: Yes but we need tests to cover them.

File: src/cljc/rems/common/application_util.cljc (hunk 10 above)
Comment (Macroz) [reply src/cljc/rems/common/application_util.cljc:51]: I'm fine with any if there are clear cases that show what the behavior is. It's a bit difficult to say if how some of these number-like parts should be sorted, and if they only ever have numbers.

Comment (Macroz) [reply src/cljc/rems/common/application_util.cljc:51]: It's quite likely that we have to amend the cases later.

Comment (Macroz) [reply src/cljc/rems/common/application_util.cljc:51]: One of the possible complications is that we have to detect which part is a year and sort by that first.

File: src/cljc/rems/common/application_util.cljc (hunk 7 above)
Comment (marharyta) [reply src/cljc/rems/common/application_util.cljc:45]: "Doesn't this also lose all the strings for comparison purposes?" - sorry @Macroz do you mean my solution?

File: CHANGELOG.md (hunk 8 above)
Comment (marharyta) [reply CHANGELOG.md:36]: right, should have formulated more specifically.

File: src/cljc/rems/common/application_util.cljc (hunk 9 above)
Comment (marharyta) [reply src/cljc/rems/common/application_util.cljc:37]: yes, I will delete it.

File: src/cljc/rems/common/application_util.cljc (hunk 7 above)
Comment (Macroz) [reply src/cljc/rems/common/application_util.cljc:45]: Any solution with `re-seq` of just digits.

File: src/cljc/rems/common/application_util.cljc [hunk 12]
```diff
@@ -27,3 +29,26 @@
...
+(defn- str-to-int [s]
+  #?(:clj (Integer/parseInt s)
+     :cljs (js/parseInt s 10)))
+
+;; (defn- printToConsole [s]
+;;   #?(:clj (println s)
+;;      :cljs (js/console.log s)))
+
+(defn parse-sortable-external-id
+  "The idea is to parse all numbers from the string to vector and then compare the values in the vector"
```
Comment (Macroz) [inline src/cljc/rems/common/application_util.cljc:42]: This is the wrong idea in that it doesn't handle now the possible differing string parts?

File: src/cljc/rems/common/application_util.cljc [hunk 13]
```diff
@@ -27,3 +29,26 @@
...
+  [external-id]
+  (when external-id
+    (when-let [number-sequence  (seq (re-seq #"\d+" external-id))]
+      (vec (map str-to-int
```
Comment (Macroz) [inline src/cljc/rems/common/application_util.cljc:46]: `mapv`

File: src/cljc/rems/common/application_util.cljc [hunk 14]
```diff
@@ -27,3 +29,26 @@
...
+                number-sequence)))))
+
+(deftest test-parse-sortable-external-id
+  (is (= [2020 10] (parse-sortable-external-id "2020/10")))
+  (is (= [54 14 01 00 2002 21] (parse-sortable-external-id "THL/54/14.01.00/2002-rems/21")))
```
Comment (Macroz) [inline src/cljc/rems/common/application_util.cljc:51]: This should not lose "THL" or "-rems" since those can be used in the comparison ...

File: src/cljc/rems/common/application_util.cljc (hunk 13 above)
Comment (marharyta) [reply src/cljc/rems/common/application_util.cljc:46]: fair enough :)
